clang/lib/Driver/ToolChains/Clang.cpp

```diff
@@ -4486,11 +4487,66 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
   if (IsIAMCU && types::isCXX(Input.getType()))
     D.Diag(diag::err_drv_clang_unsupported) << "C++ for IAMCU";
 
+  // Handle depscan.
+  if (isa<DepscanJobAction>(JA)) {
+    CmdArgs.push_back("-cc1depscan");
+
+    // Pass depscan related options to cc1depscan.
+    const OptSpecifier DepScanOpts[] = {
+        options::OPT_fdepscan_EQ,
+        options::OPT_fdepscan_share_EQ,
+        options::OPT_fdepscan_share_parent,
+        options::OPT_fdepscan_share_parent_EQ,
+        options::OPT_fno_depscan_share,
+        options::OPT_fdepscan_share_stop_EQ,
+        options::OPT_fdepscan_daemon_EQ,
+        options::OPT_fdepscan_prefix_map_EQ,
+        options::OPT_fdepscan_prefix_map_sdk_EQ,
+        options::OPT_fdepscan_prefix_map_toolchain_EQ};
+
+    Args.AddAllArgs(CmdArgs, DepScanOpts);
+
+    assert(Output.isFilename() && "Depscan needs to have file output");
+    CmdArgs.push_back("-o");
+    CmdArgs.push_back(Output.getFilename());
+    CmdArgs.push_back("-cc1-args");
+    JA = cast<DepscanJobAction>(Job).getScanningJobAction();
+  }
+
   // Invoke ourselves in -cc1 mode.
   //
   // FIXME: Implement custom jobs for internal actions.
   CmdArgs.push_back("-cc1");
 
+  // Handle full response file input.
+  if (Inputs.front().getType() == types::TY_CMD) {
+    // Render response file input first.
+    assert(Inputs.size() == 1 && "Only one response file input");
+    auto &II = Inputs.front();
+    assert(II.isFilename() && "Should be response file");
+    SmallString<128> InputName("@");
+    InputName += II.getFilename();
+    CmdArgs.push_back(C.getArgs().MakeArgString(InputName));
+
+    // FIXME: We still need to overwrite the output file path.
+    if (Output.isFilename()) {
+      CmdArgs.push_back("-o");
+      CmdArgs.push_back(Output.getFilename());
+    }
+    const char *Exec = D.getClangProgramPath();
+    if (D.CC1Main && !D.CCGenDiagnostics) {
+      // Invoke the CC1 directly in this process
+      C.addCommand(std::make_unique<CC1Command>(
+          Job, *this, ResponseFileSupport::AtFileUTF8(), Exec, CmdArgs, Inputs,
+          Output));
+    } else {
+      C.addCommand(std::make_unique<Command>(Job, *this,
+                                             ResponseFileSupport::AtFileUTF8(),
+                                             Exec, CmdArgs, Inputs, Output));
+    }
+    return;
+  }
+
```

Comment:
You should split the function here and send in the `const JobAction &JA` that you actually want to be used. Then you don't need to change all the `JA` things below.

Reply:
Or, somehow find a way to avoid changing every reference to `JA`. We'll get too many build failures and merge conflicts with this the patch looks right now.

E.g., another option is to avoid introducing the `JA` variable until you know what you want it to be set to.

Reply:
I have a better solution now. Updating.

Reply:
Nice! This looks way better.